bitsandbytes/autograd/_functions.py

```diff
@@ -312,13 +314,16 @@ def forward(ctx, A, B, out=None, bias=None, state=MatmulLtState):
             state.outlier_pool = GlobalOutlierPooler.get_instance()
 
         # Cast A to fp16
-        if A.dtype != torch.float16:
-            warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
+        A_dtype = torch.float16
```

Comment:
```suggestion
       if A.dtype != torch.float16:
            warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
            A_dtype = torch.float16
```
Tensors which are already in in fp16 do not need to be set again

Reply:
@abhilash1910 Thanks for the comment. Here we are considering other dtypes like bfloat16 for CPU.
```python
A_dtype = torch.float16
if A.device == torch.device('cpu'):
    A_dtype = torch.bfloat16
if A.dtype != A_dtype:
    warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to {A_dtype} during quantization")
```

Reply:
The conversion is done afterwards. Here is just to print a warning.
`CA, CAt, SCA, SCAt, coo_tensorA = F.double_quant(A.to(A_dtype), threshold=state.threshold)`
And in fact, if tensor is already in `A_dtype`, no action will be taken.

Reply:
Quick question that might be related here. Do we need to consider any changes (e.g. fall back to fp32) for users with a CPU that does not have AVX512-BF16 or AMX? Or is that something handled by torch?

Reply:
It will fall back to fp32 automatically. It's handled by torch.